File: 0003_longest_substring_without_repeating_characters.py
# https://leetcode-cn.com/problems/longest-substring-without-repeating-characters/

class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        Inputs:
            s: str
        Outputs:
            res: int
        """
        
        '''
        left = 0
        right = 0
        len_s = len(s)
        res = 0
        while left < len_s:
            app = []
            # Check if next string has appeared.
            while right < len_s and s[right] not in app:
                app.append(s[right])
                right = right + 1
            res = max(res, right-left)
            
            left = left + 1
            right = left
        return res
        '''
        
        # Sliding window algorithm.
        app = set() # app records strings that have appeared
        n = len(s)
        right = 0 # right pointer
        res = 0
        for i in range(n):
            if i != 0:
                app.remove(s[i-1]) # remove a string
            while right < n and s[right] not in app:
                app.add(s[right])
                right = right + 1
            
            res = max(res, right-i)
        return res

File: Cpp/0003_longest_substring_without_repeating_characters.cpp
// https://leetcode-cn.com/problems/longest-substring-without-repeating-characters/

class Solution {
public:
	int lengthOfLongestSubstring(string s) {
		unordered_map<char, int> table;
		int res = 0;
		int start = 0, end = 0;
		while (end < s.size()) {
			auto iter = table.find(s[end]);
			// s[end] already exists in table.
			if (iter != table.end()) {
				start = max(iter->second + 1, start);
				table[s[end]] = end;
			}
			// s[end] does not exist in table.
			else {
				table.emplace(s[end], end);
			}
			end = end + 1;
			res = max(res, end - start);
		}

		return res;
	}
};